bundles/org.openhab.binding.upnpcontrol/src/main/java/org/openhab/binding/upnpcontrol/internal/handler/UpnpRendererHandler.java

```diff
@@ -0,0 +1,866 @@
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.audio.AudioFormat;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.NextPreviousType;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.library.types.PercentType;
+import org.eclipse.smarthome.core.library.types.PlayPauseType;
+import org.eclipse.smarthome.core.library.types.RewindFastforwardType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.eclipse.smarthome.io.net.http.HttpUtil;
+import org.eclipse.smarthome.io.transport.upnp.UpnpIOService;
+import org.openhab.binding.upnpcontrol.internal.UpnpAudioSink;
+import org.openhab.binding.upnpcontrol.internal.UpnpAudioSinkReg;
+import org.openhab.binding.upnpcontrol.internal.UpnpEntry;
+import org.openhab.binding.upnpcontrol.internal.UpnpXMLParser;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link UpnpRendererHandler} is responsible for handling commands sent to the UPnP Renderer. It extends
+ * {@link UpnpHandler} with UPnP renderer specific logic.
+ *
+ * @author Mark Herwege - Initial contribution
+ * @author Karel Goderis - Based on UPnP logic in Sonos binding
+ */
+@NonNullByDefault
+public class UpnpRendererHandler extends UpnpHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(UpnpRendererHandler.class);
+
+    private static final int SUBSCRIPTION_DURATION_SECONDS = 3600;
+
+    // UPnP protocol pattern
+    private static final Pattern PROTOCOL_PATTERN = Pattern.compile("(?:.*):(?:.*):(.*):(?:.*)");
+
+    private volatile boolean audioSupport;
+    protected volatile Set<AudioFormat> supportedAudioFormats = new HashSet<>();
+    private volatile boolean audioSinkRegistered;
+
+    private volatile UpnpAudioSinkReg audioSinkReg;
+
+    private volatile boolean upnpSubscribed;
+
+    private static final String UPNP_CHANNEL = "Master";
+
+    private volatile OnOffType soundMute = OnOffType.OFF;
+    private volatile PercentType soundVolume = new PercentType();
+    private volatile List<String> sink = new ArrayList<>();
+
+    private volatile ArrayList<UpnpEntry> currentQueue = new ArrayList<>();
```

Comment:
Are you aware that `volatile` doesn't make it synchronized? If I see correctly, `onValueReceived()` can be invoked concurrently, so these should be synchronized.

Reply:
I am aware of this. But I am not good at it at all and always struggle to know where and how to use synchronization. I had not run into issues so far, so ignored it. I probably should not have.
onValueReceived could indeed, in theory, be called concurrently. However, each call would come from a media server and/or renderer device in sequence, so in practice, I have not seen issues so far.
How could I best proceed?

Reply:
If an object can be modified while another thread reads/writes it, the operation shall be thread-safe, although if it's very unlikely the two threads accessing the same resource concurrently. You could write the following to make it thread-safe:
```
private ArrayList<UpnpEntry> currentQueue = Collections.synchronizedList(new ArrayList<>());
```
The `volatile` can be omitted, since the reference `currentQueue` is only set during class initialization.
If you loop over the list, the whole operation shall be synchronized:
```
synchronized (currentQueue) {
   currentQueue.forEach(...);
}
```

Reply:
@fwolter Thank you for this explanation.
When I look at the specific example, currentQueue is actually not only set during class initialization. In the class initialization, it is just initialized with an empty ArrayList default. It actually is set in the registerQueue method (L696 of UpnpRendererHandler), where a new queue is passed in and completely replaces the previous one. No writes are done inside the currentQueue anywhere in the code, except full replacement of the full queue. All the rest is only reads. Do I still need synchronization for this? That's also why I think volatile is required.
I am going through the code again and looking for other potential synchronization issues.

Reply:
Your arguments are correct. Synchronization is not needed, then, and volatile is ok.